[JoanneJo] iP

JoanneJo opened the pull request.

JoanneJo committed:
Add Level 0 - Rename, Greet, Exit

JoanneJo committed:
Add Level 1 - Echo

JoanneJo committed:
Add Level 2 - Add, List

JoanneJo committed:
Add Level 3 - Mark as Done

JoanneJo committed:
Add A-CodingStandard

JoanneJo committed:
Add Level 4 - ToDos, Events, Deadlines

JoanneJo committed:
Add A-TextUiTesting

JoanneJo committed:
Add A-CodeQuality

lctxct commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,183 @@
+    public static void addNewEvent() {
+        int fromIndex = input.indexOf("/from");
+        int toIndex = input.indexOf("/to");
+        int eventWordLength = 6;
+        int fromWordLength = 6;
+        int toWordLength = 4;
```
Since these are constants, I think the names should be in uppercase?

lctxct commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,183 @@
+
+    public static void addNewDeadline() {
+        int byIndex = input.indexOf("/by");
+        int deadlineWordLength = 9;
+        int byWordLength = 4;
```
Similarly, I think these should be uppercase also since they are constants? :o

lctxct commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,183 @@
+import java.util.Scanner;
+
 public class Duke {
+    private static final Scanner IN = new Scanner(System.in);
```
The scanner is appropriately named. 👍✨

BenedictChannn replied:
Is this appropriate? 
Consider initializing your scanner object in the main class.
Is there a need to make your scanner object final?

lctxct commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,183 @@
+    private static int taskListIndex = taskCount - 1;
+
+    public static void greet() {
+        String logo = "                      .--:::.                      \n"
+                + "                .:-----::::.:-.                    \n"
+                + "            .----:::::::::..:=:                    \n"
+                + "         .---:::::::::::::::-*.                    \n"
```
Great job with the indentation down here.

lctxct reviewed (commented):
Good work with the PR overall, I had a really hard time trying to find code quality violations. I think the main thing to take note of is to uppercase constant names.

BenedictChannn commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,183 @@
+            System.out.println("Your basket is on a lemonade break right now.");
+        } else {
+            System.out.println("Your basket is looking citrusy-fresh:");
+            int taskSerialNo = 1;
```
Is this necessary?

BenedictChannn commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,183 @@
+                + "╲________╱╲________╱╲__╱__╱__╱╲________╱╲__╱_____╱  \n";
+
+        System.out.println(logo);
+        System.out.println("Hey there! I'm Lemon \uD83C\uDF4B");
```
Consider naming the unicode to provide more information.

BenedictChannn commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,183 @@
+
+    public static String getInput() {
+        System.out.println("\uD83C\uDF4B\n");
+        String line;
```
Would it be more appropriate to inline this line and line 50?

BenedictChannn reviewed (commented):
Great Job! Consider some minor changes to your code

JoanneJo committed:
Update Lemon.java

Modified the naming convention and reduced the use of global variables to fit the coding standard.

JoanneJo committed:
Add Level 5 - Handle Errors

Lemon.java: add error checking and exception
handling for todo, deadline and event tasks,
and for marking and unmarking of tasks

LemonException.java: add a class to represent
exceptions specific to Lemon

Error handling is not implemented in Lemon.java.

Wrong input from users may cause error and
exceptions causing the program to fail or exit.

Implement error and exception handling.

Implementing error handling allows the program to
respond to unexpected input gracefully. Users are
also able to know how to pass a valid input.

Error and exceptions include wrong format input,
invalid input, empty input, array index out of
bounds, null pointer, empty list, unexpected input.

JoanneJo committed:
Merge branch 'branch-Level-5'

JoanneJo committed:
Add A-Packages

Create 2 packages: lemon.command and lemon.task

LemonException.java move to lemon.command
Task.java, Todo.java, Deadline.java, Event.java
move to lemon.task

JoanneJo committed:
Merge branch 'branch-A-Packages'

Musfirahe0556596 commented on src/main/java/Lemon.java:
```diff
@@ -0,0 +1,266 @@
+
+    public static void markTask(String input, boolean isDone, boolean isEmptyList) throws LemonException {
+        if (isEmptyList) {
+            if (isDone) {
+                throw new LemonException("Your basket is on a lemonade break right now! " +
+                        "There are no tasks to be marked. Add some fruitful tasks!");
+            } else {
+                throw new LemonException("Your basket is on a lemonade break right now! " +
+                        "There are no tasks to be unmarked. Add some fruitful tasks!");
+            }
+        }
+
+        try {
+            String inputPattern = "(mark|unmark) (\\d+)";
+
+            Pattern r = Pattern.compile(inputPattern);
+            Matcher m = r.matcher(input);
+
+            if (m.find()) {
+                int taskSerialNo = Integer.parseInt(m.group(2).trim());
+                int taskIndex = taskSerialNo - 1;
+
+                if (isDone) {
+                    tasks[taskIndex].markAsDone();
+                    System.out.println("Great job! This task is now juiced:");
+                } else {
+                    tasks[taskIndex].markAsNotDone();
+                    System.out.println("No problem! This task is back into the basket:");
+                }
+
+                System.out.println("\t" + tasks[taskIndex].toString());
+            } else {
+                throw new LemonException("Oopsie! Please use the format 'mark/unmark <task number>'!");
+            }
+        } catch (NumberFormatException e) {
+            System.out.println("Oopsie! Please enter a valid task number!");
+        } catch (ArrayIndexOutOfBoundsException | NullPointerException e) {
+            System.out.println("Oopsie! Please enter a task number that is on the list!");
+        }
+    }
```
Avoid long methods that are more than 30 LoC

Musfirahe0556596 commented on src/main/java/Lemon.java:
```diff
@@ -0,0 +1,266 @@
+        System.out.print(LEMON_EMOJI + " " + LEMON_EMOJI + " "+ LEMON_EMOJI);
+    }
+
+    public static void main(String[] args) {
```
Avoid long methods that are more than 30 LoC

Musfirahe0556596 commented on src/main/java/Lemon.java:
```diff
@@ -0,0 +1,266 @@
+import lemon.task.Todo;
+
+import java.util.Scanner;
+import java.util.regex.*;
```
Imported classes should always be listed explicitly (avoid *).

Musfirahe0556596 commented on src/main/java/Lemon.java:
```diff
@@ -0,0 +1,266 @@
+        try {
+            String inputPattern = "todo (.+)";
+
+            Pattern r = Pattern.compile(inputPattern);
+            Matcher m = r.matcher(input);
```
Variables could be named better

Musfirahe0556596 commented on src/main/java/Lemon.java:
```diff
@@ -0,0 +1,266 @@
+        try {
+            String inputPattern = "deadline (.+?) /by (.+)";
+
+            Pattern r = Pattern.compile(inputPattern);
+            Matcher m = r.matcher(input);
```
Variables could be named better

Musfirahe0556596 commented on src/main/java/Lemon.java:
```diff
@@ -0,0 +1,266 @@
+        try {
+            String inputPattern = "event (.+?) /from (.+?) /to (.+)";
+
+            Pattern r = Pattern.compile(inputPattern);
+            Matcher m = r.matcher(input);
```
Variables could be named better

Musfirahe0556596 commented on src/main/java/Lemon.java:
```diff
@@ -0,0 +1,266 @@
+        try {
+            String inputPattern = "(mark|unmark) (\\d+)";
+
+            Pattern r = Pattern.compile(inputPattern);
+            Matcher m = r.matcher(input);
```
Variables could be named better

Musfirahe0556596 reviewed (commented):
Overall looks good but there are some coding quality and standards to be followed!

JoanneJo committed:
Update Lemon.java

Modified Lemon.java to fit the coding quality and
standards. Shortened markTask() method by adding
a checkEmptyList() method. Explicitly listed
imported classes and renamed variables.

JoanneJo committed:
Add Level 6 - Delete

Added the delete function in Lemon.java to
allow deletion of tasks. Exception handling was
done for this delete function.

Updated the printAddedTask method to
printAddedOrDeletedTask method to print the
status of the list for both add and delete
functions.

Updated the checkEmptyList method to send
customized exception message for the delete
function in the case of an empty list.

Changed the fixed array used to store tasks to
a resizable ArrayList. This provides dynamic
sizing and improved flexibility. Updated
various methods to replace array indexing,
retrieving, adding and deleting of objects with
ArrayList methods.

JoanneJo committed:
Add Level 7 - Save

Data will be loaded from a file in hard disk
during bootup. Tasks can now be saved in the
hard disk whenever changes are made to the
tasks. Changes include adding, deleting,
marking and unmarking of tasks. Exception and
error handling is implemented.

JoanneJo committed:
Add A-Jar

Resolved conflicts.
Fixed code for the functions to work properly.

JoanneJo committed:
Add A-MoreOOP

Created new classes Command.java, Messages.java,
Storage.java, TaskList.java, UI.java, Parser.java.

Extracted closely related functions as classes.

Command.java and *Command.java classes executes
the various commands from user.

Messages.java stores all the constant messages to
be displayed to the user.

Storage.java manages the loading and saving of
data to and from a text file.

TaskList.java manages the task list, it performs
functions like adding, deleting, marking, and
unmarking of tasks. Tasks in the task list and
size of the task list can also be retrieved here.
It can also check if the list is empty.

UI.java handles the receiving and displaying of
messages to and from the user.

Parser.java checks if the user input is valid and
of the right format before proceeding with the
execution of commands.

JoanneJo committed:
Add Level-9 Find

Added the find function for users to find tasks by
the keyword given by them.

FindCommand.java was added to execute the find
function.

ListCommand.java, Messages.java, Task.java,
TaskList.java, UI.java, Parser.java were modified
to support the find function.

The displayTaskList and displayFoundList methods
were differentiated.

JoanneJo committed:
Add A-JavaDoc

Added JavaDoc comments to all classes of the code.

JoanneJo committed:
Merge branch 'branch-Level-9'

JoanneJo committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/lemon/task/Task.java
#	src/main/java/lemon/task/TaskList.java
#	src/main/java/lemon/ui/UI.java

JoanneJo committed:
Add A-UserGuide

Added a User Guide in README.md

JoanneJo committed:
Update README.md

JoanneJo committed:
Update Messages.java
